src/ethereum_client.rs

```diff
@@ -192,56 +204,113 @@ mod tests {
         };
 
         let eth_client = EthereumClient::new(&config).expect("Unable to create Ethereum client");
 
+        let milkman = Milkman::new(config.milkman_address, Arc::clone(&eth_client.inner_client));
+
         let latest_block_num = eth_client
             .get_latest_block_number()
             .await
             .expect("Unable to get latest block number");
-
-        assert!(latest_block_num > 7994445);
-
-        let chain_timestamp = eth_client
-            .get_chain_timestamp()
+        let first_block_in_fork = latest_block_num + 1;
+        let chain_id = eth_client
+            .get_chain_id()
             .await
-            .expect("Unable to get chain timestamp");
-
-        assert!(chain_timestamp > 1669053987);
+            .expect("Unable to get chain id");
+        assert_eq!(chain_id, 1, "Test must be run on mainnet");
+        assert_eq!(
+            latest_block_num, 20920411,
+            "Test is designed for a specific mainnet block"
+        );
 
-        let goerli_uni_addr = "0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984"
-            .parse()
-            .unwrap();
-        let goerli_uni_whale = "0x41653c7d61609D856f29355E404F310Ec4142Cfb"
+        let weth = token("0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", &eth_client);
+        let dai = token("0x6B175474E89094C44Da98b954EedeAC495271d0F", &eth_client);
+        // Any EOA that has a large amount of WETH at the fork block
+        let weth_whale: H160 = "0x8eb8a3b98659cce290402893d0123abb75e3ab28"
             .parse()
             .unwrap();
 
-        let balance = eth_client
-            .get_balance_of(goerli_uni_addr, goerli_uni_whale)
+        eth_client.impersonate(&weth_whale).await.expect(
+            "Node should be able to impersonate accounts. Is the node compatible with Anvil?",
+        );
+
+        assert!(swaps_since(first_block_in_fork, &eth_client)
             .await
-            .expect("Unable to get balance");
+            .is_empty());
 
-        assert!(balance > 0.into());
+        weth.approve(milkman.address(), U256::max_value())
+            .from(weth_whale)
+            .send()
+            .await
+            .expect("Must be able to approve token");
+
+        let amount_in = u256_from_f64_saturating(1e18);
+        let receiver = H160([0x42; 20]);
+        let price_checker = H160([0x11; 20]);
+        let price_checker_data = Bytes::from([0x13, 0x37]);
+        milkman
+            .request_swap_exact_tokens_for_tokens(
+                amount_in,
+                weth.address(),
+                dai.address(),
+                receiver,
+                price_checker,
+                price_checker_data.clone(),
+            )
+            .from(weth_whale)
+            .send()
+            .await
+            .expect("Swap request must succeed");
 
-        let requested_swaps = eth_client
-            .get_requested_swaps(0, latest_block_num)
+        // Without mining an empty block, the test is flaky.
```

Comment:
Is this a matter of mining more blocks or just waiting for a short time?

Reply:
I don't recall having a block mined and not seeing the state change by inspecting the chain immediately afterwards. But the services also have quite a few code paths that can introduce delay so things like this regularly require some artificial waiting.
1. change chain state with tx
2. assume services will react to it (e.g. settle an order after injecting liquidity into a pool)
3. check on chain that user balances changed after `2`

Since `2` happens in a background tasks the test code would have to inject some waiting mechanism between `1` and `3`.
Not sure if this applies to milkman as well but it seems reasonable to me. Because as long as the reaction of milkman for something happening on-chain doesn't happen in the same block the original chain state transition happens there is a race condition.